### ISSUE TITLE
OpenSSL 3.3.2 build fails on RHEL 7

### ISSUE THREAD

**Issue by friesoft** [openssl/openssl#25366]
We get this error building OpenSSL 3.3.2 on RHEL 7. All previous OpenSSL versions (1.1.x, 3.0.x, 3.1.x, 3.2.x, 3.3.1) have been successfully built in the exact same environment:
```
+ ./config --prefix=/opt/openssl --openssldir=/opt/openssl -fPIC no-gost shared
Configuring OpenSSL version 3.3.2 for target linux-x86_64
Using os-specific seed configuration
Created configdata.pm
Running configdata.pm
Created Makefile.in
Created Makefile
Created include/openssl/configuration.h
**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL.md file first)      ***
***                                                                ***
**********************************************************************
+ make depend
+ make -j4
/usr/bin/perl util/mkinstallvars.pl PREFIX=. BINDIR=apps APPLINKDIR=ms LIBDIR= INCLUDEDIR=include "INCLUDEDIR=./include" ENGINESDIR=engines MODULESDIR=providers "VERSION=3.3.2" "LDLIBS=-ldl -pthread " > builddata.pm
/usr/bin/perl "-I." "-Iutil/perl" "-Mconfigdata" "-MOpenSSL::paramnames" "util/dofile.pl" "-oMakefile" crypto/params_idx.c.in > crypto/params_idx.c
/usr/bin/perl util/mkinstallvars.pl "PREFIX=/opt/openssl" BINDIR=bin "LIBDIR=lib64" "libdir=/opt/openssl/lib64" INCLUDEDIR=include APPLINKDIR=include/openssl "ENGINESDIR=/opt/openssl/lib64/engines-3" "MODULESDIR=/opt/openssl/lib64/ossl-modules" "PKGCONFIGDIR=/opt/openssl/lib64/pkgconfig" "CMAKECONFIGDIR=/opt/openssl/lib64/cmake/OpenSSL" "LDLIBS=-ldl -pthread " "VERSION=3.3.2" > installdata.pm
/usr/bin/perl "-I." "-Mconfigdata" "util/dofile.pl" "-oMakefile" include/crypto/bn_conf.h.in > include/crypto/bn_conf.h
"pairs" is not exported by the List::Util module
Can't continue after import errors at util/mkinstallvars.pl line 14.
BEGIN failed--compilation aborted at util/mkinstallvars.pl line 14.
make: *** [builddata.pm] Error 255
make: *** Waiting for unfinished jobs....
"pairs" is not exported by the List::Util module
Can't continue after import errors at util/mkinstallvars.pl line 14.
BEGIN failed--compilation aborted at util/mkinstallvars.pl line 14.
make: *** [installdata.pm] Error 255
```

Looks like this was already discussed here but without any outcome: https://github.com/openssl/openssl/discussions/24840

**Comment by alexdotgov2** [comment]
If you `yum install perl-List-MoreUtils` you'll get past this error. But I'm now dealing with a different error after that.

**Comment by friesoft** [comment]
Thanks for the suggestion, but nope, adding `perl-List-MoreUtils` didn't fix the error for me.

**Comment by friesoft** [comment]
Thanks for the suggestion, but nope, adding `perl-List-MoreUtils` didn't fix the error for me.

**Comment by alexdotgov2** [comment]
Sorry, you're right. The build was "proceeding" but in a broken state.

**Comment by friesoft** [comment]
Looks like the version of List::Util provided by RHEL 7 is too old:
```
[~]$ rpm -q perl-Scalar-List-Utils
perl-Scalar-List-Utils-1.27-248.el7.x86_64
```

According to the module changelog pairs were added in 1.28 (broken)/1.29: https://metacpan.org/dist/Scalar-List-Utils/changes
```
1.28 -- Thu Aug 01 12:19 UTC 2013
  -- BROKEN; do not use. See 1.29
 
  * Added pairgrep, pairmap, pairs (inspired by List::Pairwise)
```

**Comment by alexdotgov2** [comment]
It seems like this should be documented, at a minimum.

**Comment by friesoft** [comment]
Current documented minimum perl version is 5.10 with this module not listed at all (Required Perl modules):
https://github.com/openssl/openssl/blob/master/NOTES-PERL.md

The prerequisites would also need to be adapted: https://github.com/openssl/openssl/blob/master/INSTALL.md#prerequisites

Yes I agree, at a minimum this should be documented. Is there a list of supported os versions around? I couldn't find any.

**Comment by levitte** [comment]
...

darn.

`pairs` came with List::Util 1.29, but the minimum perl, 5.10, only includes List::Util 1.19.

To conform, we shouldn't have used `pairs`.  Fortunately, we already have replacement helpers elsewhere, I'll just adjust

**Comment by adamskiii** [comment]
I am also seeing the same error on Linux 7 with GCC 7

`Calling: make depend
Calling: make DIRS="crypto ssl apps" install
/usr/bin/perl util/mkinstallvars.pl PREFIX=. BINDIR=apps APPLINKDIR=ms LIBDIR= INCLUDEDIR=include "INCLUDEDIR=./include" ENGINESDIR=engines MODULESDIR=providers "VERSION=3.3.2" "LDLIBS=-ldl -pthread " > builddata.pm
"pairs" is not exported by the List::Util module
Can't continue after import errors at util/mkinstallvars.pl line 14.
BEGIN failed--compilation aborted at util/mkinstallvars.pl line 14.`

**Comment by metercai** [comment]
> Looks like the version of List::Util provided by RHEL 7 is too old:
> 
> ```
> [~]$ rpm -q perl-Scalar-List-Utils
> perl-Scalar-List-Utils-1.27-248.el7.x86_64
> ```
> 
> According to the module changelog pairs were added in 1.28 (broken)/1.29: https://metacpan.org/dist/Scalar-List-Utils/changes
> 
> ```
> 1.28 -- Thu Aug 01 12:19 UTC 2013
>   -- BROKEN; do not use. See 1.29
>  
>   * Added pairgrep, pairmap, pairs (inspired by List::Pairwise)
> ```

<img width="641"  src="https://github.com/user-attachments/assets/d31bedd8-c37f-441f-8407-07cf67f20388">

Too confused. Perl-Scalar-List-Utils-1.27 is already the latest version in RHEL7, but the Docker image quay.io/pypa/manylinux2014_x86_64 in my CI task cannot be changed, and the Rust build task cannot be rolled back to 3.3.1.

How can I upgrade to perl-Scalar-List-Utils-1.29?

**Comment by psharma-gdal** [comment]
Is there any plan for an openssl release containing this fix soon?

**Comment by psharma-gdal** [comment]
> ...
> 
> darn.
> 
> `pairs` came with List::Util 1.29, but the minimum perl, 5.10, only includes List::Util 1.19.
> 
> To conform, we shouldn't have used `pairs`. Fortunately, we already have replacement helpers elsewhere, I'll just adjust

Is there any plan for an openssl release containing this fix soon? The latest 3.3.2 release doesn't seem to have this fix. Could you please let me know that when can we expect the release having the fix?

**Comment by nhorman** [comment]
Iirc the fix using the master branch now, so it will be available in 3.4.  @levitte can you confirm please?

**Comment by t8m** [comment]
Yes, it will be available in 3.4.0.

**Comment by thespacecowboys45** [comment]
I also had this issue installing OpenSSL from the source tarball version 3.3.2 from https://openssl-library.org/source/

After running ./Configure I ran 'make' and received this error:

```

/usr/bin/perl util/mkinstallvars.pl PREFIX=. BINDIR=apps APPLINKDIR=ms LIBDIR= INCLUDEDIR=include "INCLUDEDIR=./include" ENGINESDIR=engines MODULESDIR=providers "VERSION=3.3.2" "LDLIBS=-ldl -pthread " > builddata.pm
"pairs" is not exported by the List::Util module
Can't continue after import errors at util/mkinstallvars.pl line 14.
BEGIN failed--compilation aborted at util/mkinstallvars.pl line 14.
make: *** [builddata.pm] Error 255
```

The fix (for me) was found here:

https://metacpan.org/pod/List::Util

upon clicking 'Install Instructions' on the left hand side of the page I followed the 'CPAN shell' instructions to update the List::Util module for perl to version 1.66.  These were the commands I ran:

```
perl -MCPAN -e shell
install Sub::Util

```

After this 'make' ran successfully.
